app/services/controlled_work_document_population_service.rb

```diff
@@ -0,0 +1,47 @@
+class ControlledWorkDocumentPopulationService
+  class << self
+    def call(session_data, form_type)
+      Dir.mktmpdir do |dir|
+        file_name = "#{dir}/output-form.pdf"
+        pdftk = PdfForms.new(`which pdftk`.chomp)
+        pdftk.fill_form template_path(form_type), file_name, values(session_data, form_type)
+        yield File.read(file_name).force_encoding("BINARY")
+      end
+    end
+
+    TEMPLATES = {
+      "cw1" => "lib/cw1-form.pdf",
+    }.freeze
+
+    def template_path(form_type)
+      Rails.root.join(TEMPLATES.fetch(form_type))
+    end
+
+    def values(session_data, form_type)
+      check = Check.new(session_data)
+      calculation_result = session_data["api_response"]
+      mappings = YAML.load_file(Rails.root.join("app/lib/controlled_work_mappings/#{form_type}.yml")).with_indifferent_access
```

Comment:
Nice. Am I right in thinking we can just have one YAML file with the fields that are across all forms?

Reply:
I think it depends a _lot_ on what we find when we start exploring the CW2 forms etc. If the fields all have the same names then we can have one YAML file. But if they have slightly different field names then I think we have a YML file for each of them. But that's an EL-798 problem, I think!

Reply:
Makes sense 😄 the only other thing I was thinking around the mappings is rather than YAML, maybe doing some sort of hash mapping instead like:

```
module ControlledWork
  class Mappings
    CHECKBOX = {
      Checkbox_box1: some_value,
      Checkbox_box2: some_value2,
      Checkbox_box3: some_value3,
    }.freeze

    TEXT = {
      undefined_10: some_attribute,
      undefined_11: some_attribute11,
      undefined_12: some_attribute12,
    }.freeze
  end
end

ControlledWork::Mappings:CHECKBOX.map do |pdf_key, attribute|
  some things to make it return the right output for the pdf
end
```
which might be a completely terrible idea but just wanted to throw it out there. The yml file looks good though .

Reply:
I think what you've done looks great, that yml file looks nice and clear now, and easy to amend. This looks like a good way of building the forms out to me!